### ISSUE TITLE
Improve NotificationCenter use through dependency injection

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/firefox-ios#8243]
Rather than creating a notification center each time we require it, or invoking it in the standard way

```java
NotificationCenter.default.etc
```

we should implement a protocol in some fashion similar to this:

```java
protocol Notifiable {
    var notificationCenter: NotificationCenter { get }
}

extension Notifiable {
    var notificationCenter: NotificationCenter {
        return NotificationCenter.default
    }
}
```

This would allow us to conform classes to Notifiable, and give us access to the notificationCenter easily.  More importantly, if we want to test these classes, we can create our own NotificationCenter that will allow easy testing of our functions and logic, allowing our code to be more robust and better tested. :grinning:

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-2809)

**Comment by data-sync-user** [comment]
➤ Roux Buciu commented:

The reason this wasn’t originally implemented in the Notifiable protocol was because NotificationCenter needs to be exposed to Obj-C and nothing in protocol extensions can be exposed to obj-c. If attempting to implement this in Xcode, you’ll get an error saying Non-'@objc' property 'notificationCenter' does not satisfy requirement of '@objc' protocol 'Notifiable' and if you add @obj-cc to the property, you’ll get an error saying @objc can only be used with members of classes, @objc protocols, and concrete extensions of classes.

We could do something like this:

!Screen Shot 2022-04-05 at 10.03.50 AM.png|width=1093,height=698!

But then it would only be available on classes, which is a limitation.

In other words, there’s no good workaround for this and the current implementation is best of both worlds.
